src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,577 @@
+ * {@link #combiner()}, and {@link #finisher()} must return a semantically
+ * identical result.
+ *
+ * <p>Implementations of Gatherer must not capture, retain, or expose to
+ * other threads, the references to the state instance, or the downstream
+ * {@link Downstream} for longer than the invocation duration of the method
+ * which they are passed to.
+ *
+ * <p>Performing a gathering operation with a {@code Gatherer} should produce a
+ * result equivalent to:
+ *
+ * {@snippet lang = java:
+ * Gatherer.Downstream<? super R> downstream = ...;
+ * A state = gatherer.initializer().get();
+ * for (T t : data) {
+ *     gatherer.integrator().integrate(state, t, downstream);
+ * }
+ * gatherer.finisher().accept(state, downstream);
+ * }
+ *
+ * <p>However, the library is free to partition the input, perform the
+ * integrations on the partitions, and then use the combiner function to
+ * combine the partial results to achieve a gathering operation.  (Depending
+ * on the specific gathering operation, this may perform better or worse,
+ * depending on the relative cost of the integrator and combiner functions.)
+ *
+ * <p>In addition to the predefined implementations in {@link Gatherers}, the
+ * static factory methods {@code of(...)} and {@code ofSequential(...)}
+ * can be used to construct gatherers.  For example, you could create a gatherer
+ * that implements the equivalent of
+ * {@link java.util.stream.Stream#map(java.util.function.Function)} with:
+ *
+ * {@snippet lang = java:
+ * public static <T, R> Gatherer<T, ?, R> map(Function<? super T, ? extends R> mapper) {
+ *     return Gatherer.of(
+ *         (unused, element, downstream) -> // integrator
+ *             downstream.push(mapper.apply(element))
+ *     );
+ * }
+ * }
+ *
+ * <p>Gatherers are designed to be <em>composed</em>; two or more Gatherers can
+ * be composed into a single Gatherer using the {@link #andThen(Gatherer)}
+ * method.
+ *
+ * {@snippet lang = java:
+ * // using the implementation of `map` as seen above
+ * Gatherer<Integer, ?, Integer> increment = map(i -> i + 1);
+ *
+ * Gatherer<Object, ?, String> toString = map(i -> i.toString());
+ *
+ * Gatherer<Integer, ?, String> incrementThenToString = plusOne.andThen(intToString);
+ * }
+ *
+ * As an example, in order to create a gatherer to implement a sequential
+ * Prefix Scan as a Gatherer, it could be done the following way:
+ *
+ * {@snippet lang = java:
+ * public static <T, R> Gatherer<T, ?, R> scan(
+ *     Supplier<R> initial,
+ *     BiFunction<? super R, ? super T, ? extends R> scanner) {
+ *
+ *     class State {
+ *         R current = initial.get();
+ *     }
+ *
+ *     return Gatherer.<T, State, R>ofSequential(
+ *          State::new,
+ *          Gatherer.Integrator.ofGreedy((state, element, downstream) -> {
+ *              state.current = scanner.apply(state.current, element);
+ *              return downstream.push(state.current);
+ *          })
+ *     );
+ * }
+ * }
```

Comment:
I would have appreciated a usage  example here - with a stream of input elements and what such a gatherer would produce as output.

Reply:
Wilco!